File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/ContentEntryContentCategoryJoin.java
package com.ustadmobile.lib.db.entities;

import com.ustadmobile.lib.database.annotation.UmEntity;
import com.ustadmobile.lib.database.annotation.UmIndexField;
import com.ustadmobile.lib.database.annotation.UmPrimaryKey;
import com.ustadmobile.lib.database.annotation.UmSyncLastChangedBy;
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum;
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum;

import static com.ustadmobile.lib.db.entities.ContentEntryContentCategoryJoin.TABLE_ID;


/**
 * Join entity to link ContentEntry many:many with ContentCategory
 */
@UmEntity(tableId = TABLE_ID)
public class ContentEntryContentCategoryJoin {

    public static final int TABLE_ID = 3;

    @UmPrimaryKey(autoGenerateSyncable = true)
    private long ceccjUid;

    //TODO: Migration
    @UmIndexField
    private long ceccjContentEntryUid;

    private long ceccjContentCategoryUid;

    @UmSyncLocalChangeSeqNum
    private long ceccjLocalChangeSeqNum;

    @UmSyncMasterChangeSeqNum
    private long ceccjMasterChangeSeqNum;

    @UmSyncLastChangedBy
    private int ceccjLastChangedBy;

    public long getCeccjUid() {
        return ceccjUid;
    }

    public void setCeccjUid(long ceccjUid) {
        this.ceccjUid = ceccjUid;
    }

    public long getCeccjContentEntryUid() {
        return ceccjContentEntryUid;
    }

    public void setCeccjContentEntryUid(long ceccjContentEntryUid) {
        this.ceccjContentEntryUid = ceccjContentEntryUid;
    }

    public long getCeccjContentCategoryUid() {
        return ceccjContentCategoryUid;
    }

    public void setCeccjContentCategoryUid(long ceccjContentCategoryUid) {
        this.ceccjContentCategoryUid = ceccjContentCategoryUid;
    }

    public long getCeccjLocalChangeSeqNum() {
        return ceccjLocalChangeSeqNum;
    }

    public void setCeccjLocalChangeSeqNum(long ceccjLocalChangeSeqNum) {
        this.ceccjLocalChangeSeqNum = ceccjLocalChangeSeqNum;
    }

    public long getCeccjMasterChangeSeqNum() {
        return ceccjMasterChangeSeqNum;
    }

    public void setCeccjMasterChangeSeqNum(long ceccjMasterChangeSeqNum) {
        this.ceccjMasterChangeSeqNum = ceccjMasterChangeSeqNum;
    }

    public int getCeccjLastChangedBy() {
        return ceccjLastChangedBy;
    }

    public void setCeccjLastChangedBy(int ceccjLastChangedBy) {
        this.ceccjLastChangedBy = ceccjLastChangedBy;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ContentEntryContentCategoryJoin that = (ContentEntryContentCategoryJoin) o;

        if (ceccjUid != that.ceccjUid) return false;
        if (ceccjContentEntryUid != that.ceccjContentEntryUid) return false;
        return ceccjContentCategoryUid == that.ceccjContentCategoryUid;
    }

    @Override
    public int hashCode() {
        int result = (int) (ceccjUid ^ (ceccjUid >>> 32));
        result = 31 * result + (int) (ceccjContentEntryUid ^ (ceccjContentEntryUid >>> 32));
        result = 31 * result + (int) (ceccjContentCategoryUid ^ (ceccjContentCategoryUid >>> 32));
        return result;
    }
}

File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/ContentEntryContentCategoryJoin.kt
package com.ustadmobile.lib.db.entities

import com.ustadmobile.lib.database.annotation.UmEntity
import com.ustadmobile.lib.database.annotation.UmIndexField
import com.ustadmobile.lib.database.annotation.UmPrimaryKey
import com.ustadmobile.lib.database.annotation.UmSyncLastChangedBy
import com.ustadmobile.lib.database.annotation.UmSyncLocalChangeSeqNum
import com.ustadmobile.lib.database.annotation.UmSyncMasterChangeSeqNum

import com.ustadmobile.lib.db.entities.ContentEntryContentCategoryJoin.Companion.TABLE_ID


/**
 * Join entity to link ContentEntry many:many with ContentCategory
 */
@UmEntity(tableId = TABLE_ID)
class ContentEntryContentCategoryJoin {

    @UmPrimaryKey(autoGenerateSyncable = true)
    var ceccjUid: Long = 0

    //TODO: Migration
    @UmIndexField
    var ceccjContentEntryUid: Long = 0

    var ceccjContentCategoryUid: Long = 0

    @UmSyncLocalChangeSeqNum
    var ceccjLocalChangeSeqNum: Long = 0

    @UmSyncMasterChangeSeqNum
    var ceccjMasterChangeSeqNum: Long = 0

    @UmSyncLastChangedBy
    var ceccjLastChangedBy: Int = 0

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false

        val that = o as ContentEntryContentCategoryJoin?

        if (ceccjUid != that!!.ceccjUid) return false
        return if (ceccjContentEntryUid != that.ceccjContentEntryUid) false else ceccjContentCategoryUid == that.ceccjContentCategoryUid
    }

    override fun hashCode(): Int {
        var result = (ceccjUid xor ceccjUid.ushr(32)).toInt()
        result = 31 * result + (ceccjContentEntryUid xor ceccjContentEntryUid.ushr(32)).toInt()
        result = 31 * result + (ceccjContentCategoryUid xor ceccjContentCategoryUid.ushr(32)).toInt()
        return result
    }

    companion object {

        const val TABLE_ID = 3
    }
}
